lib/voicemail-admin.js

```diff
@@ -0,0 +1,779 @@
+ * @copyright 2014, Digium, Inc.
+ * @license Apache License, Version 2.0
+ * @author Jonathan R. Rose <jrose@digium.com>
+ */
+
+'use strict';
+
+var menu = require('../menu.json');
+var prompt = require('prompt');
+
+var dal;
+var logger;
+
+var sprintf = require('sprintf').sprintf;
+
+var commands = createApi();
+
+var Q = require('q');
+
+prompt.message = '';
+prompt.delimiter = '';
+var inputIndicator = {
+  name: 'command',
+  message: '=>'.green
+};
+
+var internalCommands = ['exit', 'help'];
+
+/**
+ * Returns a promise that resolves when the requested command is finished
+ * executing.
+ *
+ * @param {object} result - command entered by prompt.get()
+ */
+function processOption(result) {
+  var entryComponents;
+  var baseCommand;
+  var chosenCommand = false;
+  var functionPtr;
+  var index;
+
+  return Q.fcall(function () {
+    if (!result || !result.command) {
+      return exit();
+    }
+
+    /* Split the string by spaces, but keep stuff between quotes together */
+    entryComponents = result.command.match(/(?:[^\s"]+|"[^"]*")+/g);
+
+    if (!entryComponents) {
+      return;
+    }
+
+    entryComponents = entryComponents.map(function(component) {
+      return component.replace(/"/g, '');
+    });
+
+    /* help is the only single word command -- everything else is two words */
+    if (entryComponents.length === 1 ||
+        internalCommands.indexOf(entryComponents[0]) > -1) {
+      baseCommand = entryComponents[0];
+    } else {
+      baseCommand = entryComponents[0] + ' ' + entryComponents[1];
+    }
+
+    chosenCommand = menu.commands[baseCommand];
+
+    if (!chosenCommand) {
+      throw new Error(sprintf('Unknown command \'%s\'', result.command));
+    }
+
+    functionPtr = commands[chosenCommand.action];
+
+    return functionPtr(chosenCommand, entryComponents);
+  })
+  .catch(function (error) {
+    logger.error('%s', error);
+  });
+}
+
+var getPrompt = Q.denodeify(prompt.get);
+
+/**
+ * Prints detailed help information for a single command to the terminal
+ *
+ * @param {array} entryComponents - all parts of the command as it was entered
+ */
+function showHelpCommand(entryComponents) {
+  var commandName;
+  var chosenCommand = false;
+
+  /* entry components must be array of 2 or greater */
+  if (entryComponents.length === 2) {
+    commandName = entryComponents[1];
+  } else {
+    commandName = entryComponents[1] + ' ' + entryComponents[2];
+  }
+
+  chosenCommand = menu.commands[commandName];
+
+  if (!chosenCommand) {
+    throw new Error('Command: \'%s\' does not exist.');
+  }
+
+  logger.info('Usage: ' + chosenCommand.usage.green);
+  logger.info(chosenCommand.description);
+
+  chosenCommand.details.forEach(function (detail) {
+    logger.info(detail);
+  });
+}
+
+/**
+ * Removes keys from an associative array that are either functions or objects
+ *
+ * @param {object} associativeArray - associative array to cull junk from
+ */
+function removeUgly(associativeArray) {
```

Comment:
Changed to 'removeNonValues' for now. Not sure if that's the best way of doing it, but eh... it seemed more correct than 'removeNonStrings'